### ISSUE TITLE
[Crash] Overscroll the BG content

### ISSUE THREAD

**Issue by Sagar0-0** [Dimezis/BlurView#223]
Please include:
1) Library version: version-2.0.5
2) Device and OS version: Pixel 8 Pro API 34 Emulator
3) Detailed steps to reproduce the issue
- Using AndroidView to Use  a blurred content in the Compose screen
- The background content is scrollable (LazyColumn similar to RecyclerView)
- If we stretch the scrollable content from top it will hit overscroll or it is what I understand.
- The app crashed because of an error attached below.

4) XML layout and code for BlurView setup
- The XML file:
```
<?xml version="1.0" encoding="utf-8"?>
<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
    android:layout_width="wrap_content"
    android:layout_height="wrap_content">

    <TextView
        android:id="@+id/first"
        android:layout_width="wrap_content"
        android:layout_height="wrap_content"
        android:layout_marginBottom="10dp"
        android:text="Some text" />

    <TextView
        android:layout_width="wrap_content"
        android:layout_height="wrap_content"
        android:layout_marginBottom="10dp"
        android:text="Some other text" />
</LinearLayout>
```
- Adding this xml as a view in the BlurView programmatically with addViewInLayout method
5) Stacktrace in case of a crash
```
java.lang.IllegalArgumentException: Software rendering doesn't support drawRenderNode
                                                                                                    	at android.graphics.Canvas.drawRenderNode(Canvas.java:2329)
                                                                                                    	at androidx.compose.foundation.DrawStretchOverscrollModifier.draw(AndroidOverscroll.android.kt:267)
                                                                                                    	at androidx.compose.ui.node.BackwardsCompatNode.draw(BackwardsCompatNode.kt:350)
                                                                                                    	at androidx.compose.ui.node.LayoutNodeDrawScope.drawDirect-eZhPAX0$ui_release(LayoutNodeDrawScope.kt:110)
                                                                                                    	at androidx.compose.ui.node.LayoutNodeDrawScope.draw-eZhPAX0$ui_release(LayoutNodeDrawScope.kt:89)
                                                                                                    	at androidx.compose.ui.node.NodeCoordinator.drawContainedDrawModifiers(NodeCoordinator.kt:450)
                                                                                                    	at androidx.compose.ui.node.NodeCoordinator.access$drawContainedDrawModifiers(NodeCoordinator.kt:58)
                                                                                                    	at androidx.compose.ui.node.NodeCoordinator$drawBlock$1$1.invoke(NodeCoordinator.kt:469)
                                                                                                    	at androidx.compose.ui.node.NodeCoordinator$drawBlock$1$1.invoke(NodeCoordinator.kt:468)
                                                                                                    	at androidx.compose.runtime.snapshots.Snapshot$Companion.observe(Snapshot.kt:503)
                                                                                                    	at androidx.compose.runtime.snapshots.SnapshotStateObserver$ObservedScopeMap.observe(SnapshotStateObserver.kt:502)
                                                                                                    	at androidx.compose.runtime.snapshots.SnapshotStateObserver.observeReads(SnapshotStateObserver.kt:258)
                                                                                                    	at androidx.compose.ui.node.OwnerSnapshotObserver.observeReads$ui_release(OwnerSnapshotObserver.kt:133)
                                                                                                    	at androidx.compose.ui.node.NodeCoordinator$drawBlock$1.invoke(NodeCoordinator.kt:468)
                                                                                                    	at androidx.compose.ui.node.NodeCoordinator$drawBlock$1.invoke(NodeCoordinator.kt:466)
                                                                                                    	at androidx.compose.ui.platform.GraphicsLayerOwnerLayer.drawLayer(GraphicsLayerOwnerLayer.android.kt:271)
                                                                                                    	at androidx.compose.ui.node.NodeCoordinator.draw(NodeCoordinator.kt:434)
                                                                                                    	at androidx.compose.ui.node.LayoutModifierNodeCoordinator.performDraw(LayoutModifierNodeCoordinator.kt:280)
                                                                                                    	at androidx.compose.ui.node.LayoutNodeDrawScope.drawContent(LayoutNodeDrawScope.kt:68)
                                                                                                    	at androidx.compose.foundation.lazy.layout.LazyLayoutItemAnimator$DisplayingDisappearingItemsNode.draw(LazyLayoutItemAnimator.kt:546)
                                                                                                    	at androidx.compose.ui.node.LayoutNodeDrawScope.drawDirect-eZhPAX0$ui_release(LayoutNodeDrawScope.kt:110)
                                                                                                    	at androidx.compose.ui.node.LayoutNodeDrawScope.performDraw(LayoutNodeDrawScope.kt:78)
                                                                                                    	at androidx.compose.ui.node.LayoutNodeDrawScope.drawContent(LayoutNodeDrawScope.kt:57)
                                                                                                    	at androidx.compose.foundation.DrawStretchOverscrollModifier.draw(AndroidOverscroll.android.kt:156)
                                                                                                    	at androidx.compose.ui.node.BackwardsCompatNode.draw(BackwardsCompatNode.kt:350)
                                                                                                    	at androidx.compose.ui.node.LayoutNodeDrawScope.drawDirect-eZhPAX0$ui_release(LayoutNodeDrawScope.kt:110)
                                                                                                    	at androidx.compose.ui.node.LayoutNodeDrawScope.draw-eZhPAX0$ui_release(LayoutNodeDrawScope.kt:89)
                                                                                                    	at androidx.compose.ui.node.NodeCoordinator.drawContainedDrawModifiers(NodeCoordinator.kt:450)
                                                                                                    	at androidx.compose.ui.node.NodeCoordinator.draw(NodeCoordinator.kt:439)
                                                                                                    	at androidx.compose.ui.node.LayoutModifierNodeCoordinator.performDraw(LayoutModifierNodeCoordinator.kt:280)
                                                                                                    	at androidx.compose.ui.node.NodeCoordinator.drawContainedDrawModifiers(NodeCoordinator.kt:447)
                                                                                                    	at androidx.compose.ui.node.NodeCoordinator.access$drawContainedDrawModifiers(NodeCoordinator.kt:58)
                                                                                                    	at androidx.compose.ui.node.NodeCoordinator$drawBlock$1$1.invoke(NodeCoordinator.kt:469)
                                                                                                    	at androidx.compose.ui.node.NodeCoordinator$drawBlock$1$1.invoke(NodeCoordinator.kt:468)
                                                                                                    	at androidx.compose.runtime.snapshots.Snapshot$Companion.observe(Snapshot.kt:2441)
                                                                                                    	at androidx.compose.runtime.snapshots.SnapshotStateObserver$ObservedScopeMap.observe(SnapshotStateObserver.kt:502)
                                                                                                    	at androidx.compose.runtime.snapshots.SnapshotStateObserver.observeReads(SnapshotStateObserver.kt:258)
                                                                                                    	at androidx.compose.ui.node.OwnerSnapshotObserver.observeReads$ui_release(OwnerSnapshotObserver.kt:133)
                                                                                                    	at androidx.compose.ui.node.NodeCoordinator$drawBlock$1.invoke(NodeCoordinator.kt:468)
                                                                                                    	at androidx.compose.ui.node.NodeCoordinator$drawBlock$1.invoke(NodeCoordinator.kt:466)
2024-10-19 12:37:02.305 11182-11182 AndroidRuntime          com.example.zomatoclone              E  	at androidx.compose.ui.platform.GraphicsLayerOwnerLayer.drawLayer(GraphicsLayerOwnerLayer.android.kt:271) (Ask Gemini)
                                                                                                    	at androidx.compose.ui.node.NodeCoordinator.draw(NodeCoordinator.kt:434)
                                                                                                    	at androidx.compose.ui.node.LayoutNode.draw$ui_release(LayoutNode.kt:1000)
                                                                                                    	at androidx.compose.ui.node.InnerNodeCoordinator.performDraw(InnerNodeCoordinator.kt:196)
                                                                                                    	at androidx.compose.ui.node.NodeCoordinator.drawContainedDrawModifiers(NodeCoordinator.kt:447)
                                                                                                    	at androidx.compose.ui.node.NodeCoordinator.draw(NodeCoordinator.kt:439)
                                                                                                    	at androidx.compose.ui.node.LayoutNode.draw$ui_release(LayoutNode.kt:1000)
                                                                                                    	at androidx.compose.ui.platform.AndroidComposeView.dispatchDraw(AndroidComposeView.android.kt:1564)
                                                                                                    	at android.view.View.draw(View.java:23892)
                                                                                                    	at android.view.View.draw(View.java:23762)
                                                                                                    	at android.view.ViewGroup.drawChild(ViewGroup.java:4556)
                                                                                                    	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4317)
                                                                                                    	at android.view.View.draw(View.java:23760)
                                                                                                    	at android.view.ViewGroup.drawChild(ViewGroup.java:4556)
                                                                                                    	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4317)
                                                                                                    	at android.view.View.draw(View.java:23892)
                                                                                                    	at eightbitlab.com.blurview.PreDrawBlurController.updateBlur(PreDrawBlurController.java:120)
                                                                                                    	at eightbitlab.com.blurview.PreDrawBlurController$1.onPreDraw(PreDrawBlurController.java:51)
                                                                                                    	at android.view.ViewTreeObserver.dispatchOnPreDraw(ViewTreeObserver.java:1176)
                                                                                                    	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:3772)
                                                                                                    	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:2465)
                                                                                                    	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:9305)
                                                                                                    	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1339)
                                                                                                    	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1348)
                                                                                                    	at android.view.Choreographer.doCallbacks(Choreographer.java:952)
                                                                                                    	at android.view.Choreographer.doFrame(Choreographer.java:882)
                                                                                                    	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1322)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:958)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:99)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:205)
                                                                                                    	at android.os.Looper.loop(Looper.java:294)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:8177)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:971)
```
> I am creating this for the sake of creating a Blur Compose Library, Everything works fine, but kust few issue, like this overscroll crash and lags if BG is GIF

**Comment by Sagar0-0** [comment]
[Screen_recording_20241019_124506.webm](https://github.com/user-attachments/assets/f3ca5248-6824-4bb2-99b7-4d00eb29f261)

**Comment by Dimezis** [comment]
This library does not support Compose.
While I could work around this particular crash by just try-catching the view snapshotting, it would most likely result in a missing blur for the time of the overscroll (or other event that can cause the same problem)

**Comment by Sagar0-0** [comment]
I know this library does not directly support Compose, that's why I am creating a Compose Library(wrapper around your BlurView) to support compose. 
- Yes, can you please fix this issue and release a new version or is there a way I can do something as a quick fix.
- another issue is Bavkground Lag in case of GIF, I can create another ticket for it once this is done. Everything else works smoohly for Jetpack Compose Project.

**Comment by Dimezis** [comment]
> that's why I am creating a Compose Library(wrapper around your BlurView) to support compose.

This isn't going to work well. 
Compose rendering relies on hardware acceleration in too many places. BlurView uses software rendering to snapshot the Compose hierarchy, and you'll just have an endless pile of bugs cropping up, even if you aren't stumbling on them now.

**Comment by Sagar0-0** [comment]
You might be right here, and because of this Rendering difference, I am getting lag/desync with GIF backgrounds. Do you have any resources that I can follow to dive deeper into Rendering and maybe achieve this in Compose?

**Comment by Dimezis** [comment]
There's already a good Compose lib for that - [Haze](https://github.com/chrisbanes/haze)
It works starting with API 31, but you won't be able to get around that anyway, because you need the RenderNode API.

**Comment by Sagar0-0** [comment]
I know about Haze but I need more supported versions below, there can be some ways, afaik Compose does use Software rendering in some [cases](https://www.reddit.com/r/android_devs/comments/1fgkvrg/comment/lnjvqb0/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button). 

Do you know, if we can make something with the existing API? or this isn't the right place for this...

**Comment by Dimezis** [comment]
🙂 
That link proves nothing, it's just a random person spreading some BS.
Text rendering has been [hardware accelerated](https://medium.com/@romainguy/androids-font-renderer-c368bbde87d9) for a long while.

Also, it doesn't matter whether Compose falls back to software rendering in _some_ cases.

Long story short, with the current state of Compose, you won't be able to pull this off **reliably** using software snapshotting. With hardware snapshotting, your only reasonable option is either RenderNode API, or TextureView/SurfaceView in combination with [lockHardwareCanvas](https://developer.android.com/reference/android/view/SurfaceHolder#lockHardwareCanvas()) to obtain a hardware canvas.
The latter would require OpenGL/Vulkan to perform blur efficiently, and it's a significantly different approach overall.

**Comment by Sagar0-0** [comment]
If anyone wants to explore jetpack compose. Here's a [library ](https://github.com/desugar-64/imla)that supports backward compatiability.

**Comment by Sagar0-0** [comment]
@Dimezis  can you please fix this issue by using try catch in the expected position or I can also do it...

**Comment by Dimezis** [comment]
> make PR 

☝️ this.

I'm not making and releasing a new version without testing how it behaves at least in your case.
Fork it, add your changes, see what's going on during the overscroll, and decide whether it's worth it.

**Comment by Sagar0-0** [comment]
assign it to me, i'll soon start working on it,

**Comment by Dimezis** [comment]
The Compose team [fixed](https://issuetracker.google.com/issues/380783671) this crash. Waiting for the release

**Comment by Dimezis** [comment]
The fix was released in Compose 1.8.0. Closing the issue
